docs/rfcs/2024-08-06-json-datatype.md

```diff
@@ -0,0 +1,141 @@
+---
+Feature Name: Json Datatype
+Tracking Issue: https://github.com/GreptimeTeam/greptimedb/issues/4230
+Date: 2024-8-6
+Author: "Yuhan Wang <profsyb@gmail.com>"
+---
+
+# Summary
+This RFC proposes a method for storing and querying JSON data in the database.
+
+# Motivation
+JSON is widely used across various scenarios. Direct support for writing and querying JSON can significantly enhance the database's flexibility.
+
+# Details
+
+## User Interface
+The feature introduces a new data type for the database, similar to the common JSON type. Data is written as JSON strings and can be queried using functions.
+
+For example:
+```SQL
+CREATE TABLE IF NOT EXISTS test (
+    ts TIMESTAMP TIME INDEX,
+    a INT,
+    b JSON
+);
+
+INSERT INTO test VALUES(
+    0,
+    0,
+    '{
+        "name": "jHl2oDDnPc1i2OzlP5Y",
+        "timestamp": "2024-07-25T04:33:11.369386Z",
+        "attributes": { "event_attributes": 48.28667 }
+    }'
+);
+
+SELECT json_get(b, 'name') FROM test;
++---------------------+
+| b.name              |
++---------------------+
+| jHl2oDDnPc1i2OzlP5Y |
++---------------------+
+
+SELECT json_get_by_paths(b, 'attributes', 'event_attributes') + 1 FROM test;
++-------------------------------+
+| b.attributes.event_attributes |
++-------------------------------+
+|                      49.28667 |
++-------------------------------+
+
+```
+
+## Storage
+
+### Schema Inference
+Unlike other types, the schema of JSON data is inconsistent. For different JSON columns, we introduce a dynamic schema inference method for storing the data.
+
+For example:
+```JSON
+{
+  "a": "jHl2oDDnPc1i2OzlP5Y",
+  "b": "2024-07-25T04:33:11.369386Z",
+  "c": { "d": 48.28648 }
+}
+```
+This will be parsed at runtime and stored as a corresponding `Struct` type in Arrow:
+```Rust
+Struct(
+    Field("a", Utf8),
+    Field("b", Utf8),
+    Field("c", Struct(Field("d", Float64))),
+)
+```
+
+Dynamic schema inference helps achieve compression in some scenarios. See [benchmark](https://github.com/CookiePieWw/json-format-in-parquet-benchmark/) for more information.
+
+## Schema Change
+The schema must remain consistent for a column within a table. When inserting data with different schemas, schema changes may occur. There are two types of schema changes:
```

Comment:
Will json schema change trigger a flush?

Reply:
Yes, it should trigger a flush.

Reply:
> Yes, it should trigger a flush.

I think there is **no actual schema change**. The actual type (parquet column schema) is inferred from the data in the memtable and the column schema is only known in the parquet files.  i.e., 

// t1, field a, parquet column schema: object { a: string }
// t2, field a, parquet column schema: object { a: string, b: int  }
// t3, field a, parquet column schema: object { a: object { c: string }  }

For the upper components, mito, or frontend component. They only know the field `a` is a JSON datatype column.

Reply:
Ah I see, so the schemas can keep different in the memtable, and when flushing into parquet files, we merge different schemas. It's more like an implicit compaction instead of an explicit change.